cctools-llvm: init at 11.1.0-973.0.1

reckenrode opened the pull request:
Note: This is targeting staging because it will be needed by Darwin stdenv changes that will need to go through staging.

###### Description of changes

cctools-llvm is a replacement for cctools that replaces as much of cctools with equivalents from LLVM that it can reasonably do. This was motivated by wanting to reduce dependencies on cctools, which are updated infrequently by upstream.

To provide a motivating example, the version of `strip` included in cctools cannot properly strip the archives in compiler-rt in LLVM 15. Paths are left to bootstrap tools, resulting in failed requisites checks in the final stdenv build. Since `strip` needs replaced, the opportunity was taken to replace other provided they are functional replacements.

Note: This has to be done in cctools (or some equivalent) because some derivations (noteably LLVM) use the bintools of the stdenv directly instead of going through the wrapper.

The following tools from LLVM are not used in this derivation:

* LLD - not fully compatible with ld64 yet and potentially too big of a change;
* libtool - not a drop-in replacement yet because it does not support linker passthrough, which is needed by xcbuild;
* lipo - crashes when running the LLVM test suite;
* install_name_tool - fails when trying to build swift-corefoundation; and.
* randlib - not completely a drop-in replacement, so leaving it out for now.

If other incompatabilities are found, the tools can be reverted or made conditional. For example, cctools `strip` is preferred on older versions of LLVM (which lack the compiler-rt issue) or when cctools itself is a new enough version because `llvm-strip` on LLVM 11 produces files that older verions of `codesign_allocate` cannot process correctly.

One final caveat/note: Some tools are not duplicated or linked from cctools-port. The names of the tools and which ones were linked was determined based on what is provided upstream in Xcode and is installed on macOS system.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

reckenrode committed:
cctools-llvm: init at 11.1.0-973.0.1

cctools-llvm is a replacement for cctools that replaces as much of cctools with equivalents from LLVM that it can reasonably do. This was motivated by wanting to reduce dependencies on cctools, which are updated infrequently by upstream.

To provide a motivating example, the version of `strip` included in cctools cannot properly strip the archives in compiler-rt in LLVM 15. Paths are left to bootstrap tools, resulting in failed requisites checks in the final stdenv build. Since `strip` needs replaced, the opportunity was taken to replace other provided they are functional replacements.

Note: This has to be done in cctools (or some equivalent) because some derivations (noteably LLVM) use the bintools of the stdenv directly instead of going through the wrapper.

The following tools from LLVM are not used in this derivation:

* LLD - not fully compatible with ld64 yet and potentially too big of a change;
* libtool - not a drop-in replacement yet because it does not support linker passthrough, which is needed by xcbuild;
* lipo - crashes when running the LLVM test suite;
* install_name_tool - fails when trying to build swift-corefoundation; and.
* randlib - not completely a drop-in replacement, so leaving it out for now.

If other incompatabilities are found, the tools can be reverted or made conditional. For example, cctools `strip` is preferred on older versions of LLVM (which lack the compiler-rt issue) or when cctools itself is a new enough version because `llvm-strip` on LLVM 11 produces files that older verions of `codesign_allocate` cannot process correctly.

One final caveat/note: Some tools are not duplicated or linked from cctools-port. The names of the tools and which ones were linked was determined based on what is provided upstream in Xcode and is installed on macOS system.

toonn commented on pkgs/os-specific/darwin/cctools/llvm.nix:
```diff
@@ -0,0 +1,89 @@
+    ln -s "${lib.getBin llvmPackages.bintools-unwrapped}/bin/llvm-ar" "$out/bin/${targetPrefix}ar"
+
+    for tool in ${toString llvm_bins}; do
+      llvmTool=''${tool/_/-}
```
May be just me and a minor nitpick, so feel free to ignore. I feel like the translation would be a bit clearer the other way around. Have `llvm_bins` list the names that are actually included with LLVM and map to the cctools name we're aliasing it as?

reckenrode replied:
I’ve made the swap in my branch.

reckenrode replied:
This is swapped in the latest push.

toonn commented on pkgs/os-specific/darwin/cctools/llvm.nix:
```diff
@@ -0,0 +1,89 @@
+    mkdir -p "$out/bin"
+
+    ln -s "${lib.getDev cctools-port}" "$dev"
+    ln -s "${lib.getMan cctools-port}" "$man"
```
Won't this be somewhat misleading? I believe you've mentioned flags that don't correspond for `ranlib`?

reckenrode replied:
Libsystem expects to find `include/mach-o` in cctools.dev. I can modify Libsystem to explicitly use `cctools-port.dev` instead to get the headers if that makes more sense. Eventually, it should use cctools-apple, but we’d need to update and make it actually build (which the new version of `ld64` doesn’t even on x86_64-darwin).

For the man pages, I can symlink them from their sources.

reckenrode replied:
I made linking the manpages optional, enabled by default but disabled in the stdenv. Linking LLVM’s manpages requires building them, which pulls in a whole bunch of Python stuff.

toonn replied:
The `dev` part isn`t a problem. It's just taking the cctools man pages even though the binaries are from LLVM that seems like it could potentially cause confusion.

reckenrode replied:
The current version of the commit addresses the man pages issue. It takes the cctools ones from cctools and the LLVM ones from LLVM.

I made whether to include them conditional because the LLVM man pages require a bunch of Python dependencies. For the stdenv build, they aren’t built because they’re not included in the output. If a user builds the man pages, they’ll get the mix appropriate to the tools in darwin.cctools-port.out.

toonn reviewed (commented):
Looks pretty good to me though I don't think we should target staging. The cycles aren't usually longer than a couple of days so I don't think that'll unnecessarily delay things.

IIUC this won't actually be used until a later PR, right?

reckenrode commented on pkgs/os-specific/darwin/cctools/llvm.nix:
```diff
@@ -0,0 +1,89 @@
+    ln -s "${lib.getDev cctools-port}" "$dev"
+    ln -s "${lib.getMan cctools-port}" "$man"
+
+    # Use the clang-integrated assembler instead of using `as` from cctools.
```
That’s actually copied from `darwin.binutils`, which has been using the integrated assembler on aarch64-darwin for a couple of years now (see #115167). Per that PR, Apple also uses the integrated assembler, so using it over the one in cctools would better match what Apple does in their toolchain.

`llvm-ar` appears to support the same flags and modifiers. I’ve not had to patch around it unlike `ranlib`.

reckenrode commented on pkgs/os-specific/darwin/cctools/llvm.nix:
```diff
@@ -0,0 +1,89 @@
+in
+stdenv.mkDerivation {
+  pname = "cctools-llvm-${if useLLD then "lld" else "ld64-${cctoolsVersion}"}";
+  version = llvmVersion;
```
Something like this?

```suggestion
  version = llvmVersion + lib.optionalString (!useLLD) "-${cctoolsVersion}";
```

reckenrode replied:
The name would then be cctools-llvm-lld or cctools-llvm-ld64 depending on which (though the LLD support is mostly theoretical at this point).

reckenrode replied:
Since I dropped the LLD stuff, this is simplified now.

toonn replied:
This comment wasn't about the LLD part of the version, just the package version, but looking at your changes you seem to have settled on what I meant any way : )

reckenrode commented:
I’m doing a rebuild to make sure it works, but I’ll be pushing an update to this PR later tonight that incorporates your feedback @toonn. I’ve also made a few additional changes:

* I dropped `useLLD` because it’s untested and complicates things; and
* Reverted back to `ranlib` from cctools because `llvm-ranlib` is not a drop-in replacement.

reckenrode commented:
> IIUC this won't actually be used until a later PR, right?

Correct. Once the prerequisite PRs are merged, I’ll open a PR with the stdenv rework that decouples it from the bootstrap tools. It will still target LLVM 11, but that will set things up to do a separate PR that only bumps LLVM.

toonn commented on pkgs/os-specific/darwin/cctools/llvm.nix:
```diff
@@ -0,0 +1,101 @@
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  # The `man` output has to be included unconditionally because dariwn.binutils expects it.
```
```suggestion
  # The `man` output has to be included unconditionally because darwin.binutils expects it.
```

reckenrode replied:
I fixed the typo and repushed.

toonn commented on pkgs/os-specific/darwin/cctools/llvm.nix:
```diff
@@ -0,0 +1,101 @@
+
+    for tool in ${toString cctools_bins}; do
+      ln -s "${lib.getBin cctools-port}/bin/${targetPrefix}$tool" "$out/bin/${targetPrefix}$tool"
+      ${lib.optionalString enableManpages ''linkManPage "${lib.getMan cctools-port}" "$tool"''}
```
I understand making the LLVM man pages optional but maybe the cctools man pages don't need to be. Those don't need to be built, right?

Or do you consider them not all that useful?

reckenrode replied:
There’s no harm in including them unconditionally. I was just trying to keep the semantics of `enableManpages` consistent by having it include either all or no man pages.

Let me know if I should make that change, though it wouldn’t affect users since building the man pages is the default. It’s only the stdenv that won’t build them (because it doesn’t need them).

toonn reviewed (commented):
Trying to clarify what I meant regarding the man pages.

reckenrode commented:
The latest push fixes the issue with `darwin.cctools` not being the same thing as `darwin.cctools-port`.

toonn reviewed (commented):
Looks like case-insensitive HFS strikes again for `cctools-apple`:

```
trying https://opensource.apple.com/tarballs/ld64/ld64-609.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
                                 Dload  Upload   Total   Spent    Left  Speed
100   257  100   257    0     0   2776      0 --:--:-- --:--:-- --:--:--  2793
100   257  100   257    0     0   2394      0 --:--:-- --:--:-- --:--:--  2401
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
100  754k    0  754k    0     0   355k      0 --:--:--  0:00:02 --:--:-- 1253k
error: hash mismatch in fixed-output derivation '/nix/store/529jabac9d4h16ppwchichndhhg93z07-ld64-609.tar.gz.drv':
         specified: sha256-SqQ7SqmK+uOPijzxOTqtkEu3qYmcth6H7rrQ03R1Q+4=
            got:    sha256-1idktrbJAsDhZcJj3t4aXbYGAGFKr0FKiNtamdUtqiU=
```

And `cctools-llvm` seems to be missing `machocheck`, I don't think that was intentional?

```
/nix/store/nwrx6x29jpn2s4g4bfqymrqxfy6x7kl4-cctools-llvm-11.1.0-973.0.1/bin/machocheck: sh: /nix/store/nwrx6x29jpn2s4g4bfqymrqxfy6x7kl4-cctools-llvm-11.1.0-973.0.1/bin/machocheck: No such file or directory
```

reckenrode commented:
> Looks like case-insensitive HFS strikes again for `cctools-apple`

I’m using case-sensitive APFS, and I get the same error. I think the hashes changed when Apple started redirecting downloads to GitHub.

> And cctools-llvm seems to be missing machocheck, I don't think that was intentional?

I only linked binaries that correspond to what Apple ships (as part of macOS or Xcode), so an omission may be intentional. However, I don’t see `macrocheck` in cctools-port, and grepping the source (and nixpkgs) turns up nothing. I’m not sure what that is.

toonn commented:
@reckenrode, not macrocheck, `machocheck`.

toonn reviewed (commented):
Still getting different hashes. I'll try GC'ing tomorrow after the configd build gets a gcroot but I think Nix wouldn't give different paths nonetheless if `cctools` and `cctools-port` were the same thing?

```
toonn@terra ~/s/n/r/pr234859-cctools-llvm> nix-build --no-out-link --pure -A dar
win.cctools
/nix/store/k3aflp9mdq8ks8pzslj9sz1jml9si6rq-cctools-port-973.0.1
toonn@terra ~/s/n/r/pr234859-cctools-llvm> nix-build --no-out-link --pure -A dar
win.cctools-port
/nix/store/b6dlp7jb98wn5fggshbcksipyyxp7gjc-cctools-port-973.0.1
```

reckenrode commented:
I created #235975 to fix the hashes on cctools-apple. I’ll also be opening a separate PR to fix cctools-port on x86_64-darwin when built with clang 16.

reckenrode commented:
> @reckenrode, not macrocheck, `machocheck`.

Replying here to document since this was discussed on Matrix: Apple doesn’t appear to ship `machocheck`, so it’s not included. It looks like it’s [used][1] by ld64 for some tests.

[1]: https://github.com/search?q=org%3Aapple-oss-distributions+machocheck&type=code

reckenrode commented:
I think I fixed the `cctool` ≠ `cctools-port` issue. I’m not sure using `rec` was the right approach, but neither `self.cctools-port` nor `pkgs.targetPackages.darwin.cctools-port` is the same as the cctools-port that’s created directly by `callPackage`.

toonn reviewed (commented):
I'm still getting different paths for them even with the recursive attribute set and I can't explain why. I've GC'ed so there's no FOD shenanigans involved.

```
> git log -1 --oneline; nix-build --pure --no-out-link -A darwin.cctools -A darwin.cctools-port; sed -n '63p;103,115p' pkgs/top-level/darwin-packages.nix
bc81c565e6c5 (HEAD, reckenrode/cctools-llvm) cctools-llvm: init at 11.1.0-973.0.1
/nix/store/k3aflp9mdq8ks8pzslj9sz1jml9si6rq-cctools-port-973.0.1
/nix/store/b6dlp7jb98wn5fggshbcksipyyxp7gjc-cctools-port-973.0.1
impure-cmds // appleSourcePackages // chooseLibs // rec {
  cctools = cctools-port;

  cctools-apple = callPackage ../os-specific/darwin/cctools/apple.nix {
    stdenv = if stdenv.isDarwin then stdenv else pkgs.libcxxStdenv;
  };

  cctools-llvm = callPackage ../os-specific/darwin/cctools/llvm.nix {
    stdenv = if stdenv.isDarwin then stdenv else pkgs.libcxxStdenv;
  };

  cctools-port = callPackage ../os-specific/darwin/cctools/port.nix {
    stdenv = if stdenv.isDarwin then stdenv else pkgs.libcxxStdenv;
  };
```

reckenrode commented:
@toonn I figured out the problem. The stdenv defines an overlay that sets darwin.cctools to the one built during the stdenv build. That results in darwin.cctools having a different value from darwin.cctools-port. I didn’t catch that during development because my stdenv rework branch was already including the additional attributes. When I cherry-picked the commit onto master, I didn’t update the stdenv to include cctools-port. I pushed an update that does.

reckenrode commented:
```sh
$ nix-build --pure --no-out-link -A darwin.cctools -A darwin.cctools-port
/nix/store/3p2xdd5xmrc0qyi10lf95nbydcyprqy9-cctools-port-973.0.1
/nix/store/3p2xdd5xmrc0qyi10lf95nbydcyprqy9-cctools-port-973.0.1
```

toonn commented on pkgs/stdenv/darwin/default.nix:
```diff
@@ -651,7 +651,7 @@ rec {
           inherit (darwin) dyld ICU Libsystem Csu libiconv rewrite-tbd;
         } // lib.optionalAttrs (super.stdenv.targetPlatform == localSystem) {
-          inherit (darwin) binutils binutils-unwrapped cctools;
+          inherit (darwin) binutils binutils-unwrapped cctools cctools-port;
```
```suggestion
          inherit (darwin) binutils binutils-unwrapped cctools-port;
```

Looks like we can drop cctools because it's just `cctools-port` anyway.

reckenrode replied:
I think we’d want the override to make sure darwin.cctools is the same as what’s used in the stdenv. Dropping cctools by itself should be okay though.

toonn replied:
Reasons in favor of keeping the override:

- One fewer build of the same stuff

Reasons in favor of dropping the override:

- Reduces complexity of the `stdenv/default.nix`

reckenrode replied:
I pushed a commit that drops `cctools` and left `cctools-port`.

```fish
$ nix-build --pure --no-out-link -A darwin.cctools -A darwin.cctools-port
/nix/store/3p2xdd5xmrc0qyi10lf95nbydcyprqy9-cctools-port-973.0.1
/nix/store/3p2xdd5xmrc0qyi10lf95nbydcyprqy9-cctools-port-973.0.1
```

toonn reviewed (commented):
What's the rationale for keeping the override BTW? We could also drop both altogether, right?

reckenrode commented:
The latest push fixed a typo. Even with man pages enabled, it wasn’t installing any of them because the check for `$sourcePath` was missing the leading `$`.

toonn reviewed (commented):
Hmm, am I wrong in expecting this to give me the man pages?

```
> nix-build --no-out-link --pure -E 'let ps = (import ./. {}); in ps.lib.getMan ps.darwin.cctools-llvm'
/nix/store/s8vk4mzjzq6q06fja0rd4ainb9ck0wbi-cctools-llvm-11.1.0-973.0.1-man
> ls -a /nix/store/s8vk4mzjzq6q06fja0rd4ainb9ck0wbi-cctools-llvm-11.1.0-973.0.1-man
./  ../
```

reckenrode commented:
> Hmm, am I wrong in expecting this to give me the man pages?

No, you’re not wrong. Are you using the latest commits? I pushed some fixes last night when I found that enabling the man pages in my rework branch wasn’t working as expected.

toonn reviewed (commented):
One more quandary. The symlink for the `otool` man page isn't working for some reason, even though the target works fine. Can you reproduce this?

```
> ls -l /nix/store/2jk0qgn8xwiqcm5pd6zjf9aq5sgdqryc-cctools-llvm-11.1.0-973.0.1-man/share/man/man1/otool.1.gz 
lrwxr-xr-x  1 root  wheel  94  1 jan  1970 /nix/store/2jk0qgn8xwiqcm5pd6zjf9aq5sgdqryc-cctools-llvm-11.1.0-973.0.1-man/share/man/man1/otool.1.gz@ -> /nix/store/vscadj4dfp9vd6ymhw9y3nk0a8ca1nkf-cctools-port-973.0.1-man/share/man/man1/otool.1.gz
> man /nix/store/vscadj4dfp9vd6ymhw9y3nk0a8ca1nkf-cctools-port-973.0.1-man/share/man/man1/otool.1.gz
> man /nix/store/2jk0qgn8xwiqcm5pd6zjf9aq5sgdqryc-cctools-llvm-11.1.0-973.0.1-man/share/man/man1/otool.1.gz
fopen: No such file or directory
Cannot open man page /nix/store/2jk0qgn8xwiqcm5pd6zjf9aq5sgdqryc-cctools-llvm-11.1.0-973.0.1-man/share/man/man1/llvm-otool.1.gz
No manual entry for /nix/store/2jk0qgn8xwiqcm5pd6zjf9aq5sgdqryc-cctools-llvm-11.1.0-973.0.1-man/share/man/man1/otool.1.gz
```

Another slightly puzzling thing is even though there's a symlink for the `ranlib` man page, `man` opens the man page for `libtool`, the contents of the pages seems like it may be equivalent but the files differ according to diff. Not sure whether there's a way to get the exact page to show up? (This is mostly out of interest, a single redundant man page shouldn't stop this PR.)

reckenrode commented:
I see what’s happening with the man pages. The man page for `otool` just includes `llvm-otool.1`, which does not exist in the cctools-llvm man output. I can special case it. However, it still won’t load because the man page is compressed. You can see an example of that below for cctools-port. It seems like that should be addressed in cctools. I’ve pushed an updated branch that special cases `otool` and `ld64`.

```
/usr/bin/man -d  /nix/store/6q78w47d2y74p97x52ps5xg7f9d6c9g1-cctools-port-973.0.1-man/share/man/man1/otool.1.gz
-- Using pager: less -R
-- Using manual sections: 1:8:2:3:3lua:n:4:5:6:7:9:l
-- Searching PATH for man directories
--   Adding /etc/profiles/per-user/reckenrode/share/man to manpath
--   Adding /run/current-system/sw/share/man to manpath
--   Adding /usr/share/man to manpath
--   Adding /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/share/man to manpath
--   Adding /Applications/Xcode.app/Contents/Developer/usr/share/man to manpath
--   Adding /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/share/man to manpath
-- Adding default manpath entries
-- Parsing config file: /etc/man.conf
-- Using manual path: /etc/profiles/per-user/reckenrode/share/man:/run/current-system/sw/share/man:/usr/share/man:/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/share/man:/Applications/Xcode.app/Contents/Developer/usr/share/man:/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/share/man
-- Using locale paths: en_US.UTF-8:en.UTF-8:.
-- Using non-standard page width: 185
-- Searching for /nix/store/6q78w47d2y74p97x52ps5xg7f9d6c9g1-cctools-port-973.0.1-man/share/man/man1/otool.1.gz
-- Found a usable page, displaying that
-- /nix/store/6q78w47d2y74p97x52ps5xg7f9d6c9g1-cctools-port-973.0.1-man/share/man/man1/otool.1.gz includes man1/llvm-otool.1
--   Unable to find man1/llvm-otool.1
```

toonn reviewed (approved):
LGTM!
